---
title: "Proteomic"
author: Alexis Hucteau
date: "`r Sys.Date()`"
output:
    html_document:
      toc: yes
      toc_float: yes
      theme: united
editor_options:
  markdown:
    wrap: sentence
---


```{r}
rm(list = ls())

suppressPackageStartupMessages({
  library(dplyr)
  library(FactoMineR)
  library(factoextra)
  library(sva)
  library(xlsx)
  library(clusterProfiler)
  library(pheatmap)
  library(rdist)
  library(DEP)
  library(SummarizedExperiment)
})

deal_with_NA <- function(df, column, value_to_change, final_value){
  for (i in column){
    print(i)
    data.table::set(df,which(df[[i]] == value_to_change), i, final_value)
  }
}

"%ni%" <- Negate("%in%")
```

```{r}
data.dir <- "/media/alexis/DATA/Proteomic/"

Discovery_Cohort_Proteomic <- read.csv(paste0(data.dir,"Discovery_cohort.csv"))

Discovery_Cohort_Proteomic_unimputed <- read.csv(paste0(data.dir,"Discovery_cohort_unimputed.csv"), check.names = F)

Cell_lines_Proteomic <- read.csv(paste0(data.dir,"Cell_lines.csv"))

CD34_Proteomic <- read.csv(paste0(data.dir,"CD34.csv"), check.names = F)

Validation_cohort_Proteomic <- read.csv(paste0(data.dir,"Validation_cohort.csv"), dec = ",")
```

```{r}
Discovery_clinicals <- readxl::read_excel("/media/alexis/DATA/Proteomic/1-s2.0-S1535610822000587-mmc7/Data_S1_Discovery_Cohort.xlsx", sheet = "Clinical Characteristics")

Discovery_Genotype <- readxl::read_excel("/media/alexis/DATA/Proteomic/1-s2.0-S1535610822000587-mmc7/Data_S1_Discovery_Cohort.xlsx", sheet = "PanelSeq I")

Discovery_Genotype_II <- readxl::read_excel("/media/alexis/DATA/Proteomic/1-s2.0-S1535610822000587-mmc7/Data_S1_Discovery_Cohort.xlsx", sheet = "PanelSeq II")

Discovery_Genotype_II <- Discovery_Genotype_II[,c(1,3,4,2)]
colnames(Discovery_Genotype_II) <- c("Pat_ID", "Type", "Frequency", "Gene")

Discovery_Genotype <- rbind(Discovery_Genotype, Discovery_Genotype_II)
rm(Discovery_Genotype_II)
```



```{r}
Make_mutation_subgroups <- function(mutations_of_interest, mutations_to_skip, genotype = Discovery_Genotype, no_mutated_patients = c("F126", "F86", "F129", "F30")){
  genotype_of_interest <- dplyr::filter(genotype, Gene %in% mutations_of_interest) %>% .$Pat_ID %>% unique()
  patient_to_skip <- dplyr::filter(genotype, Gene %in% mutations_to_skip) %>% .$Pat_ID %>% unique()
  genotype_control <- genotype$Pat_ID[genotype$Pat_ID %ni% patient_to_skip] %>% unique()
  genotype_of_interest <- intersect(genotype_of_interest, genotype_control)
  genotype_control <- c(genotype_control[genotype_control %ni% genotype_of_interest], no_mutated_patients)
  
  pheno <- data.frame(Patient_ID = c(genotype_of_interest, patient_to_skip, genotype_control), 
                      pheno = c(rep("Mutations_of_interest", length(genotype_of_interest)), rep("Others", length(patient_to_skip)), rep("Group_control", length(genotype_control))))
  return(pheno)  
}
```



```{r}
Pheno <- Make_mutation_subgroups(c("IDH1", "IDH2"), c("DNMT3A", "TET2", "DNMT1", "NPM1"))
rownames(Pheno) <- Pheno$Patient_ID

Pheno <- Pheno[colnames(Discovery_Cohort_Proteomic)[1:177],]
```


```{r}
data <- Discovery_Cohort_Proteomic_unimputed
data$PG.Genes %>% duplicated() %>% any()

data %>% group_by(PG.Genes) %>% summarize(frequency = n()) %>% arrange(desc(frequency)) %>% filter(frequency > 1)
data_unique <- make_unique(data, "PG.Genes", "PG.UniProtIds", delim = ";")

data_unique_log_reverted <- sapply(data_unique[,1:177], function(x){
  sapply(x, function(y){
    2**y
  })
})

data_unique_log_reverted <- cbind(data_unique_log_reverted, data_unique[,177:183])
pheno_se <- data.frame("label" = colnames(data)[1:177], condition = Pheno$pheno, replicate = 1:177)

data_se <- make_se(data_unique_log_reverted, 1:177, pheno_se)
data_se
plot_frequency(data_se)
```

```{r}
data_filt <- filter_missval(data_se, thr = 0)
data_filt2 <- filter_missval(data_se, thr = 1)
plot_numbers(data_filt)
```

```{r}
plot_coverage(data_filt)
```

```{r}
for (i in colnames(data_filt@assays@data@listData[[1]])){
    data_filt@assays@data@listData[[1]][,i][is.nan(data_filt@assays@data@listData[[1]][,i])]<-NA
}

data_norm <- normalize_vsn(data_filt)

plot_normalization(data_filt[,1:5], data_norm[,1:5])
```

```{r}
plot_missval(data_filt[, 1:10])
```

```{r}
plot_detect(data_filt)
```

```{r}
data_imp <- impute(data_filt, fun = "MinProb", q = 0.01)
data_imp_man <- impute(data_filt, fun = "man", shift = 1.8, scale = 0.3)

plot_imputation(data_filt, data_imp)
plot_imputation(data_filt, data_imp_man)
```

```{r}
data_diff <- test_diff(data_imp, type = "control", control = "Group_control")

dep <- add_rejections(data_diff, alpha = 0.05, lfc = log2(1.5))
```

```{r}
plot_pca(dep, n = 500, point_size = 4, label = F, indicate = "condition")
```

```{r}
plot_cor(dep, significant = FALSE, lower = 0, upper = 1, pal = "Reds")
```

```{r}
plot_heatmap(dep, type = "centered", kmeans = FALSE, 
             k = 3, show_row_names = FALSE,
             indicate = "condition")
```

```{r}
plot_heatmap(dep, type = "contrast", kmeans = FALSE, 
             k = 6, col_limit = 10, show_row_names = FALSE)
```

```{r}
plot_volcano(dep, contrast = "Mutations_of_interest_vs_Group_control", label_size = 2, add_names = TRUE, adjusted = F)
```

```{r}
plot_single(dep, proteins = "IDH1", type = "centered") + theme(legend.position = "none")
```

```{r}
# plot_cond(dep)
```

```{r}
df_long <- get_df_long(dep)
```

##### CD34 data 

```{r}
colnames(CD34_Proteomic)[1:13] <- stringr::str_remove(colnames(CD34_Proteomic)[1:13], pattern = "20210617_EXPL2_AsJa_SA_Revision_Healthy_DIAproteome_CD34cells_") %>% stringr::str_remove(pattern = ".raw.PG.Quantity") %>% stringr::str_split(pattern = "] ") %>% lapply(function(x) x[2]) %>% unlist()
```

```{r}
Merged_patient_and_CD34 <- merge(Discovery_Cohort_Proteomic_unimputed, CD34_Proteomic, by.x = "PG.UniProtIds", by.y = "UniProtIds", all.x =T, all.y = T)

data_merged <- Merged_patient_and_CD34

data_merged$Genes %>% duplicated() %>% any()

data_merged %>% group_by(Genes) %>% summarize(frequency = n()) %>% arrange(desc(frequency)) %>% filter(frequency > 1)
data_merged_unique <- make_unique(data_merged, "PG.Genes", "PG.UniProtIds", delim = ";")

data_merged_unique_log_reverted <- data_merged_unique
data_merged_unique_log_reverted[,c(2:178, 182:194)] <- sapply(data_merged_unique_log_reverted[,c(2:178, 182:194)], function(x){
  sapply(x, function(y){
    2**y
  })
})

Group_control <- Pheno[Pheno$pheno=="Group_control",]
Mutants <- Pheno[Pheno$pheno=="Mutations_of_interest",]
Others <- Pheno[Pheno$pheno=="Others",]

Group_control$replicat <- 1:nrow(Group_control)
Mutants$replicat <- 1:nrow(Mutants)
Others$replicat <- 1:nrow(Others)

CD34_pheno <- data.frame("Patient_ID" = colnames(data_merged)[182:194], "pheno" = rep("CD34", 13), "replicat" = 1:13)

pheno_merged_se <- rbind(Group_control, Mutants, Others, CD34_pheno)
rownames(pheno_merged_se) <- pheno_merged_se$Patient_ID
pheno_merged_se <- pheno_merged_se[colnames(data_merged)[c(2:178, 182:194)],]
colnames(pheno_merged_se) <- c("label", "condition", "replicate")

data_for_batch <- data_merged_unique_log_reverted
for (i in colnames(data_for_batch[,c(2:178, 182:194)])){
    data_for_batch[,c(2:178, 182:194)][,i][is.nan(data_for_batch[,c(2:178, 182:194)][,i])]<-NA
}

batch <- ifelse(pheno_merged_se$condition == "CD34", "CD34", "Patients")

data_for_batch <- dplyr::filter(data_for_batch, !is.na(PG.Genes) & !is.na(Genes))

data_for_batch <- data.frame(data_for_batch[rowSums(is.na(data_for_batch[,2:178])) < 176 & rowSums(is.na(data_for_batch[,182:194])) < 12,])
Batch_removed_data <- sva::ComBat(as.matrix(data_for_batch[,c(2:178, 182:194)]), batch)
Batch_removed_data <- cbind(Batch_removed_data, data_for_batch[,c(1, 179:181, 195:200)])
data_merged_batched_se <- make_se(Batch_removed_data, 1:190, pheno_merged_se)

data_merged_batched_se <- make_se(data_for_batch, c(2:178, 182:194), pheno_merged_se)
data_merged_batched_se
plot_frequency(data_merged_batched_se)
```

```{r}
data_merged_batched_filt <- filter_missval(data_merged_batched_se, thr = 0)
data_merged_batched_filt2 <- filter_missval(data_merged_batched_se, thr = 1)
plot_numbers(data_merged_batched_filt)
```

```{r}
plot_coverage(data_merged_batched_filt) + theme(legend.position = "none")
```

```{r}
data_merged_batched_norm <- normalize_vsn(data_merged_batched_filt)
```

```{r}
plot_detect(data_merged_batched_filt)
```

```{r}
data_merged_batched_imp <- impute(data_merged_batched_filt, fun = "MinProb", q = 0.01)
data_merged_batched_imp_man <- impute(data_merged_batched_filt, fun = "man", shift = 1.8, scale = 0.3)

plot_imputation(data_merged_batched_filt, data_merged_batched_imp)
plot_imputation(data_merged_batched_filt, data_merged_batched_imp_man)
```

```{r}
data_merged_batched_diff <- test_diff(data_merged_batched_imp, type = "control", control = "Group_control")

dep_merged <- add_rejections(data_merged_batched_diff, alpha = 0.05, lfc = log2(1.5))
```

```{r}
plot_pca(dep_merged, n = 500, point_size = 4, label = F, indicate = "condition")
```

```{r}
plot_cor(dep_merged, significant = TRUE, lower = 0, upper = 1, pal = "Reds")
```

```{r}
plot_heatmap(dep_merged, type = "centered", kmeans = TRUE, 
             k = 2, show_row_names = FALSE,
             indicate = "condition")
```

```{r}
plot_heatmap(dep_merged, type = "contrast", kmeans = FALSE, 
             k = 6, col_limit = 10, show_row_names = FALSE)
```

```{r}
plot_volcano(dep_merged, contrast = "Mutations_of_interest_vs_Group_control", label_size = 2, add_names = TRUE, adjusted = F)
```

```{r}
plot_single(dep_merged, proteins = "IDH1", type = "centered") + theme(legend.position = "none")
```

```{r}
plot_cond(dep_merged)
```

```{r}
Diff_Prot_exp <- dep_merged@elementMetadata@listData %>% as.data.frame()

write.table(Diff_Prot_exp, "Results/Proteo/Diff_Prot_exp.tsv", quote = F, row.names = F, sep = "\t")
```

```{r}
fPPI <- read.table("Results/Proteo/fPPI_network.tsv", sep = "\t", header = T)
Diff_Prot_exp <- read.table("Results/Proteo/Diff_Prot_exp.tsv", sep = "\t", header = T) %>% as.data.frame()
Diff_Prot_exp <- Diff_Prot_exp %>% 
  dplyr::filter(Mutations_of_interest_vs_Group_control_p.val < 0.1) %>% 
  dplyr::select(name, Mutations_of_interest_vs_Group_control_diff, Mutations_of_interest_vs_Group_control_p.val, ProteinDescriptions)

Diff_Prot_exp <- Diff_Prot_exp[order(Diff_Prot_exp$Mutations_of_interest_vs_Group_control_diff, decreasing =T),]
write.table(Diff_Prot_exp, "Results/Proteo/Diff_Prot_exp_filtered.tsv", sep = "\t", quote = F, row.names = F)

fPPI <- dplyr::filter(fPPI, source %in% Diff_Prot_exp$name | target %in% Diff_Prot_exp$name)
write.table(fPPI, "Results/Proteo/fPPI_network_filtered.tsv", sep = "\t", quote = F, row.names = F)
```

